designs/full-cluster-lifecycle-api-cloudstack.md

```diff
@@ -0,0 +1,85 @@
+
+### Defaults
+
+As part of the CLI logic, we set default values to some spec if they're missing from what customers provide, like [datacenter availability zone](https://github.com/aws/eks-anywhere/blob/ed4425dadb19600b4eb446d29b81f5c2441c16f6/pkg/api/v1alpha1/cloudstackdatacenterconfig_types.go#L216), which can be leveraged in mutation webhook. 
+
+We don't want to modify the spec that customers have already specified. [Control lPlane Host Port](https://github.com/aws/eks-anywhere/blob/3c1fd0ff732641ed02137213863942403f59c320/pkg/providers/cloudstack/validator.go#L211) could be missing from what customers provide in control plane host, and the reconciler would set a default port when generating CAPI objects. 
+
+Cloudstack machine config doesn’t have default values at this stage.
+
+### Data validation
+
+In order to better sort out data validation and run time validation, we’ll reorganize validations in CLI.
+
+* Extract data validation from ValidateClusterMachineConfigs (https://github.com/aws/eks-anywhere/blob/main/pkg/providers/cloudstack/validator.go#L127) and called from webhook.
+* We have immutable fields validation in both webhook ([validateImmutableFieldsCloudStackMachineConfig](https://github.com/aws/eks-anywhere/blob/ed4425dadb19600b4eb446d29b81f5c2441c16f6/pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go#L86) and provider [validateMachineConfigImmutability](https://github.com/aws/eks-anywhere/blob/01cd1e7c3da0c6d87b2d85c4ac6e61f409091e9d/pkg/providers/cloudstack/cloudstack.go#L162)). There’re some gaps between two places so we’ll need to decide the final immutable fields.
+  * affinity group
+  * [disk offering](https://github.com/aws/eks-anywhere/issues/5319)
+* We have [k8s version](https://github.com/aws/eks-anywhere/blob/ed4425dadb19600b4eb446d29b81f5c2441c16f6/pkg/providers/cloudstack/cloudstack.go#L1371) limitation due to capc version. This lightweight validation can be done in cluster webhook.
+
+### Runtime Validation
+
+#### Datacenter Reconciler
+
+We’ll have CloudStackDatacenterReconciler to validate and update corresponding status and failure message.
+
+#### How to read credential
+
+  The datacenter reconciler needs a validator with cloudstack credential to build cmk to talk to CloudStack API. The reconciler will retrieve the CloudStack credentials from secrets. These secrets are created by the CLI during the management cluster creation or added by customers. Cloudstack supports multiple secrets by referring secret name and credential profile name.
+
+  CLI doesn't allow customers to [rotate credentials](https://github.com/aws/eks-anywhere/blob/main/designs/cloudstack-multiple-endpoints.md?plain=1#L187). Since reconciler reads credentials from secrets dynamically, customers are allowed to rotate secrets.
```

Comment:
Why don't we change the CLI as well?

Reply:
There's a ticket in FLC backlog on how to handle secrets overall https://github.com/aws/eks-anywhere-internal/issues/1176